mopidy/internal/storage.py

```diff
@@ -0,0 +1,60 @@
+from __future__ import absolute_import, unicode_literals
+
+import gzip
+import json
+import logging
+import os
+import tempfile
+
+from mopidy import models
+from mopidy.internal import encoding
+
+logger = logging.getLogger(__name__)
+
+
+def load(path):
+    """
+    Deserialize data from file.
+
+    :param path: full path to import file
+    :type path: str
+    :return: deserialized data
+    :rtype: dict
+    """
+    # Todo: raise an exception in case of error?
+    if not os.path.isfile(path):
+        logger.info('File does not exist: %s.', path)
+        return {}
+    try:
+        with gzip.open(path, 'rb') as fp:
+            return json.load(fp, object_hook=models.model_json_decoder)
+    except (IOError, ValueError) as error:
+        logger.warning(
+            'Loading JSON failed: %s',
+            encoding.locale_decode(error))
+        return {}
+
+
+def save(path, data):
+    """
+    Serialize data to file.
+
+    :param path: full path to export file
+    :type path: str
```

Comment:
My comment was a tad too brief, and didn't explain why at all.

Most of the Python stdlib handles both bytestring and unicode paths. However, on the disk, the directory and file names are stored as paths. Usually, they are all encoded from text to bytes using the same encoding, which is hopefully known to Python. Sometimes the file system encoding is unknown or there are some dirs and files that has byte sequences in their names that are not decodable using the encoding Python believes the file system is using. In those cases, converting the path to unicode throws away information and it is not possible to reconstruct a bytestring path from the unicode path that can be used to read the file. You simply no longer know the exact path to the file.

In Mopidy, we've used bytestrings for paths as consistently to avoid this (arguably quite small) problem, and thus new Mopidy APIs should continue this practice. URIs, however, are unicode strings as they are URI-encoded and thus only contains ASCII chars. URI-encoded unicode strings contain exact information about arbitrary bytes, and can be converted back to the exact same bytestring.